### PR TITLE
Enable kubecontainer.Cache in kubelet

### PR DESCRIPTION
**Pull request by yujuhong**
Depends on #17700

### CODE REVIEW

**Comment by yujuhong** [comment]
Only the last commit cddd507b7b4e5e587401530c4933d9a13516d6fb is new

**Comment by Random-Liu** [comment]
Will review soon, :0

**Comment by Random-Liu** [comment]
LGTM with small nits and some questions.

**Comment by Random-Liu** [comment]
Please rebase to fix the unit test, :0

**Comment by yujuhong** [comment]
@Random-Liu, I'm not too keen on rebasing right now, given that I'll have to do it again soon :P

**Comment by Random-Liu** [comment]
@yujuhong Hahahaha, I see.

File: pkg/kubelet/pod_workers.go [hunk 1]
```diff
@@ -99,13 +102,27 @@ func newPodWorkers(runtimeCache kubecontainer.RuntimeCache, syncPodFn syncPodFnT
...
 	}
 }
 
 func (p *podWorkers) managePodLoop(podUpdates <-chan workUpdate) {
 	var minRuntimeCacheTime time.Time
+
 	for newWork := range podUpdates {
 		err := func() (err error) {
+			podID := newWork.pod.UID
+			if p.podCache != nil {
```
Comment (yifan-gu) [inline pkg/kubelet/pod_workers.go]: Can we create a fake cache, so we can remove all the `if p.podCache != nil` ?

Comment (yifan-gu) [inline pkg/kubelet/pod_workers.go]: @yujuhong If you want to do that in follow up PR, it's ok.

**Comment by yifan-gu** [comment]
LGTM.

File: pkg/kubelet/pod_workers.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/pod_workers.go]: Let's keep it for a while. I want to keep an easy way to bypass this cache for now. I will clean them up later.

**Comment by Random-Liu** [comment]
LGTM, hope we could remove those wired code soon. :)

**Comment by timothysc** [comment]
@yujuhong , @jeremyeder - cpu performance is drastically different with this patch "in a good way".

**Comment by timothysc** [comment]
One thing I notice now appears to be startup races:  3m		3m		1	{kubelet 127.0.0.1}	spec.containers{hello}	Warning		Failed		Failed to start container with docker id 44ddf93c6304 with error: API error (500): Cannot start container 44ddf93c63042516db8da1fa93eba837d50a5add1f41137097f59df45fe2cf47: [8] System error: open /sys/fs/cgroup/cpu,cpuacct/system.slice/docker-44ddf93c63042516db8da1fa93eba837d50a5add1f41137097f59df45fe2cf47.scope/cpu.shares: no such file or directory

  2m	2m	1	{kubelet 127.0.0.1}		Warning	FailedSync	Error syncing pod, skipping: not all containers have started: 0 != 1

**Comment by piosz** [comment]
This PR broke kubernetes-e2e-gce. IIUC this is the second try to merge it. Please make sure the tests are passing before trying to merge it for the 3rd time.

cc @dchen1107

**Comment by timothysc** [comment]
@piosz do you have details on how it broke?

**Comment by yujuhong** [comment]
> This PR broke kubernetes-e2e-gce. IIUC this is the second try to merge it. Please make sure the tests are passing before trying to merge it for the 3rd time.

@piosz, this PR was only merged once, and there was no "second try" as you mentioned. I'll take a look at the builds to see how it may have broken the tests.

**Comment by piosz** [comment]
@timothysc In general Reboot tests became more flaky after the PR was merged (I didn't investigate the problem deeper). Between runs 10388 (when PR was merged) and 10415 (when PR was reverted) the suite failed 14/26 times. After the PR was reverted the suite failed 3/18 times.

@yujuhong sorry for that, I should have misunderstood something.

**Comment by yujuhong** [comment]
I looked at two failed builds (kubernetes-e2e-gce/10414 and 10406) and found a potential bug that led to Reboot test failures. This potential bug is not related to this PR, so I am not sure yet why the PR may have caused the bug the surface more often. 


After rebooting the node, the test check if all the nodes are ready *and* all the pods on the nodes (in the kube-system namespace) are running and ready.
```
05:57:41 Jan 18 05:57:41.683: INFO: Pod "fluentd-elasticsearch-e2e-gce-minion-mouw" in namespace "kube-system" disappeared. Error: pods "fluentd-elasticsearch-e2e-gce-minion-mouw" not found
05:57:41 Jan 18 05:57:41.683: INFO: Pod kube-proxy-e2e-gce-minion-mouw failed to be running and ready.
05:57:41 Jan 18 05:57:41.684: INFO: Wanted all 2 pods to be running and ready. Result: false. Pods: [kube-proxy-e2e-gce-minion-mouw fluentd-elasticsearch-e2e-gce-minion-mouw]
05:57:53 Jan 18 05:57:53.678: INFO: Node e2e-gce-minion-mouw failed reboot test.
```
From the log snippet, node `e2e-gce-minion-mouw` failed because pod `fluentd-elasticsearch-e2e-gce-minion-mouw` disappeared from etcd. Note that `fluentd-elasticsearch-e2e-gce-minion-mouw` is a mirror pod created to reflect the status of its counterpart, the static pod from a pod manifest file on the node.

Let's check `kubelet.log` to see what happens around that time.
```
I0118 13:57:41.605108    2519 kubelet.go:1634] Failed creating a mirror pod for "fluentd-elasticsearch-e2e-gce-minion-mouw_kube-system(3d933132a83d2b5b74d473a70bfee97f)": pods "fluentd-elasticsearch-e2e-gce-minion-mouw" already exists
I0118 13:57:41.618280    2519 manager.go:371] Status for pod "fluentd-elasticsearch-e2e-gce-minion-mouw_kube-system(8c4b9941-bde1-11e5-9e6d-42010af00002)" updated successfully: {status:{Phase:Running Conditions:[{Type:Ready Status:False...
I0118 13:57:41.620072    2519 kubelet.go:2255] SyncLoop (RECONCILE, "api"): "fluentd-elasticsearch-e2e-gce-minion-mouw_kube-system(8c4b9941-bde1-11e5-9e6d-42010af00002)"
I0118 13:57:41.626722    2519 kubelet.go:2249] SyncLoop (UPDATE, "api"): "fluentd-elasticsearch-e2e-gce-minion-mouw_kube-system(8c4b9941-bde1-11e5-9e6d-42010af00002)"
I0118 13:57:41.629266    2519 manager.go:382] Pod "fluentd-elasticsearch-e2e-gce-minion-mouw_kube-system(8c4b9941-bde1-11e5-9e6d-42010af00002)" fully terminated and removed from etcd
```
kubelet sends a status update the this pod, and then deletes the pod from the etcd. Why would kubelet delete the pod?
From the code, kubelet [deletes a pod](https://github.com/kubernetes/kubernetes/blob/0051dcc83e0847615c4e7cab99480fc2d662d1fd/pkg/kubelet/status/manager.go#L382) after a status update if the [deletion timestamp of the pod is set](https://github.com/kubernetes/kubernetes/blob/0051dcc83e0847615c4e7cab99480fc2d662d1fd/pkg/kubelet/status/manager.go#L374) and the [containers are not running](https://github.com/kubernetes/kubernetes/blob/0051dcc83e0847615c4e7cab99480fc2d662d1fd/pkg/kubelet/status/manager.go#L377). This is how kubelet deletes a pod regularly. The only question is why the deletion timestamp is set for the pod. Even though kubelet creates/deletes mirror pods, it uses non-graceful deletion and from the log, kubelet did not try to delete this specific mirror pod at all.

To understand why the deletion timestamp was set for the pod, I checked the `apiserver.log` for non-GET operations against this pod.
```
I0118 12:48:24.882906       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (2.783904ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:37393]
I0118 12:51:15.715561       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (6.416964ms) 200 [[kube-controller-manager/v1.2.0 (linux/amd64) kubernetes/1534672/node-controller] 127.0.0.1:44237]
I0118 12:53:07.420623       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (2.596768ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:37798]
I0118 13:28:17.880718       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (2.631787ms) 200 [[kube-controller-manager/v1.2.0 (linux/amd64) kubernetes/1534672/node-controller] 127.0.0.1:48535]
I0118 13:32:41.927229       6 handlers.go:130] DELETE /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw: (34.531092ms) 200 [[kube-controller-manager/v1.2.0 (linux/amd64) kubernetes/1534672/node-controller] 127.0.0.1:48698]
I0118 13:33:03.577484       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (3.061695ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:39978]
I0118 13:33:07.358951       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (3.527757ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:39981]
I0118 13:33:17.367898       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (3.406235ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:39989]
I0118 13:42:03.909704       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (2.701706ms) 200 [[kube-controller-manager/v1.2.0 (linux/amd64) kubernetes/1534672/node-controller] 127.0.0.1:49757]
I0118 13:43:57.276981       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (2.915222ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:41995]
I0118 13:57:41.607339       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (3.566871ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:47035]
I0118 13:57:41.628846       6 handlers.go:130] DELETE /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw: (7.747719ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:47036]
I0118 13:57:46.780298       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (3.559384ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:47050]
I0118 13:59:22.538431       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (4.388346ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:45321]
I0118 13:59:28.244047       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (3.331595ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:45333]
I0118 14:05:39.485897       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (2.912964ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:34979]
I0118 14:05:45.098770       6 handlers.go:130] PUT /api/v1/namespaces/kube-system/pods/fluentd-elasticsearch-e2e-gce-minion-mouw/status: (2.580712ms) 200 [[kubelet/v1.2.0 (linux/amd64) kubernetes/1534672] 10.240.0.5:34990]
```

At 13:32:41.927229  (***25 min before the test failure***), the node controller tried deleting the pod in question. This was during another reboot test: `Reboot [Disruptive] each node by dropping all outbound packets for a while and ensure they function afterwards`. 
```
I0118 13:32:41.891340       6 nodecontroller.go:829] Starting deletion of pod fluentd-elasticsearch-e2e-gce-minion-mouw
I0118 13:32:41.891725       6 event.go:210] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"fluentd-elasticsearch-e2e-gce-minion-mouw", UID:"8c4b9941-bde1-11e5-9e6d-42010af00002", APIVersion:"v1", ResourceVersion:"6344", FieldPath:""}): type: 'Normal' reason: 'NodeControllerEviction' Marking for deletion Pod fluentd-elasticsearch-e2e-gce-minion-mouw from Node e2e-gce-minion-mouw
```
The node controller uses graceful deletion and only forcefully remove the pods after the grace period has expired. Because kubelet treats the static pod as the source of the truth, it will not try to delete the pods/containers when seeing the deletion timestamp being set in the mirror pod. Later, the node became ready and the pod eviction was cancelled:
```
I0118 13:33:00.704344       6 nodecontroller.go:486] Node e2e-gce-minion-mouw is ready again, cancelled pod eviction
```
This means that the mirror pod in question stays in `Terminating` (`DeletionTimestamp` is set), yet `Running` status after the reboot test at 13:32:41, eventually leading to the deletion of the mirror pod in a much later test.

Summary:

- Kubelet should not delete the mirror pod even if the deletion timestamp is set. Note that we should probably make mirror pods read-only to avoid issues like this, but that's an on-going discussion (see #16627)
- Node controller should either not try to evict static pods and/or forcefully terminating the pods with deletion timestamp set already. cc @kubernetes/goog-control-plane 
- The failed test was affected by another test which completed 25 minutes ago. We should do a better job isolating the tests, e.g., making sure no pod is in the Terminating state.

I will check other failed builds to see if they show the same symptom.

**Comment by timothysc** [comment]
nice sleuthing @yujuhong

**Comment by yujuhong** [comment]
There is another cause with the same symptom (missing mirror pod) in kubernetes-e2e-gce-reboot/10683. This one is caused by kubelet incorrectly deleting and recreating the mirror pod. The test checked the pod during the recreation process and failed. I upload a fix for this: #19838.

**Comment by yujuhong** [comment]
1. Build failures caused by https://github.com/kubernetes/kubernetes/pull/19436#issuecomment-172982445: kubernetes-e2e-gce/10409, 10411, 10414
2. Build failures caused by https://github.com/kubernetes/kubernetes/pull/19436#issuecomment-173024084: kubernetes-e2e-gce/10418, 10432, kubernetes-e2e-gce-reboot/10683

So far the builds I checked failed due to either (1) or (2). None of them are directly caused by this reverted PR. I'll update the logic in kubelet, and file issues for other potential improvements.

**Comment by gmarek** [comment]
@yujuhong - I completely agree with the third bullet, I don't have any strong opinion on first one, but I'm not quite sure about the second. Is there a way to check if given pod is a mirror pod? Aren't they exactly the same as ordinary one from the control plane's perspective?

**Comment by yujuhong** [comment]
> @yujuhong - I completely agree with the third bullet, I don't have any strong opinion on first one, but I'm not quite sure about the second. Is there a way to check if given pod is a mirror pod? Aren't they exactly the same as ordinary one from the control plane's perspective?

The only difference is that there is an annotation key `kubernetes.io/config.mirror` set in the mirror pods. Whether or not apiserver should distinguish mirror pods from regular pods is another issue. I have been told to not make mirror pods special from the beginning, but it has caused us/users confusion and issues...I'll leave that discussion to #16627.

**Pull request merged**